README.md
# Flutter PubNub Plugin

A cross platform plugin for implementing real time messaging applications using PubNub (https://www.pubnub.com).
This plugin is not provided nor supported by PubNub. It was implemented part of a real time messaging application developed by Ingenio (https://www.ingenio.com).

## Features

* Multi client support allowing to handle messaging and presence from multiple PubNub clients defined in the PubNub environment.
* Auth key handling
* Filter expression handling
* Presence capabilities
* Publish and Subscribe/Unsubscribe capabilities
* Channel Groups
* History
* Push notifications
* Status 
* Error handling
* Message Actions

**IMPORTANT NOTES**:

* The plugin can only be used on iOS and Android and not supported on Flutter Web. In order to support Web, Desktop and mobile platforms, PubNub would have to create a pure Dart plugin that directly accesses their low level REST API.

## Roadmap

The plugin does not cover all functionalities offered by PubNub but is functional for implementing already complex real time messaging apps on iOS and Android.
We have items we will be working on soon:

PNConfig features:
- cipher key
- custom origin

## Getting Started

The plugin follows as much as possible the naming conventions PubNub exposes via their respective iOS and Android SDK.

Note: version 0.2.0 had a bug related to missing proguard rules, causing crashes when the library was initialized.

## Easy setup

```dart
import 'package:flutter_pubnub/pubnub.dart';


final PubNub _client = PubNub(PubNubConfig('pub-c-xxxx', 'sub-cxxx'));

```

The PubNubConfig constructor allows to pass a PubNub auth key, presence timeout in seconds, a filter expression and finally a UUID (note a default UUID is generated if not provided)

```dart

PubNubConfig(this.publishKey, this.subscribeKey, {this.authKey, this.presenceTimeout, this.uuid, this.filter});

```

## Table of contents

* [Creating one or more clients](#Creating-one-or-more-clients)
* [Subscribe to channels and unsubscribe from channels](#Subscribe-to-channels-and-unsubscribe-from-channels)
* [Publishing to channels](#Publishing-to-channels)
* [Handle presence](#Handle-presence)
* [Subscribe to PubNub events](#Subscribe-to-PubNub-Events)
* [Channel Groups](#Channel-Groups)
* [Retrieve history](#Retrieve-history)
* [Push Notifications](#Push-Notifications)
* [Signals](#Signals)
* [Message Actions](#Message-actions)
* [Cleanup](#Cleanup)

## Creating one or more clients

Creating one or more clients with PubNub subscribe and publish keys

```dart

// Note that the proper key values for PubNub are found in the client configuration dashboard under your PubNub account
final PubNub _client1 = PubNub(PubNubConfig('pub-c-xxxx', 'sub-cxxx'));
final PubNub _client2 = PubNub(PubNubConfig('pub-c-yyyy', 'sub-cyyy'));

```

Creating a client with a PubNub auth key

```dart

// Note that the proper key values for PubNub are found in the client configuration dashboard under your PubNub account
final PubNub _client = PubNub(PubNubConfig('pub-c-xxxx', 'sub-cxxx', authKey: 'auth-xxxx'));

```

Creating a client with a UUID

```dart

// Note that the proper key values for PubNub are found in the client configuration dashboard under your PubNub account
final PubNub _client = PubNub(PubNubConfig('pub-c-xxxx', 'sub-cxxx', uuid: '127c1ab5-fc7f-4c46-8460-3207b6782007'));

```

Creating a client with a presence timeout in seconds

```dart

// Note that the proper key values for PubNub are found in the client configuration dashboard under your PubNub account
final PubNub _client = PubNub(PubNubConfig('pub-c-xxxx', 'sub-cxxx', presenceTimeout: 120));

```

Creating a client with a filter expression

```dart

// The filter expression here is done, as an example on the uuid, so any messages sent from the client with that UUID will not be received in the subscribe method
final PubNub _client = PubNub(PubNubConfig('pub-c-xxxx', 'sub-cxxx', filter: 'uuid != "127c1ab5-fc7f-4c46-8460-3207b6782007"'));

```

Then in order to publish a message with the proper metadata that will trigger the filtering

```dart

// The filter expression here is done, as an example on the uuid, so any messages sent from the client with that UUID will not be received in the subscribe method
_client.publish(['Test-Channel1'], {'message': 'Hello World'},
               metadata: {
                'uuid': '127c1ab5-fc7f-4c46-8460-3207b6782007'
                }
              );

```

## Retrieve the client UUID

In some cases, especially when the UUID is not setup during client creation, it can be helpful getting the UUID that was automatically generated by PubNub.

```dart

_client.uuid().then((uuid) => print('UUID: $uuid'));

```

## Subscribe to channels and unsubscribe from channels

A client can subscribe to one or many channels. It can also unsubscribe from one or many/all channels

```dart

 _client.subscribe(['Test-Channel1', 'Test-Channel1']);

 _client.unsubscribe(['Test-Channel1']);
 
 _client.unsubscribeAll();

```


## Publishing to channels

A client can publish messages to one or many channels

```dart

 _client.publish((['Test-Channel1'], {'message': 'Hello World!'});

```

The publish returns a Future<Map> with message related items such as uuid, affected channels, timeToken, ...

```dart

 _client.publish((['Test-Channel1'], {'message': 'Hello World!'}).then((result) => print('Time Token: ${result['timeToken']}'));

```

As previously described, if a filter has been set and if we need to react to it, metadata related to the configured filter must be passed during publication.

```dart

_client.publish(['Test-Channel1'], {'message': 'Hello World'},
               metadata: {
                'uuid': '127c1ab5-fc7f-4c46-8460-3207b6782007'
                }
              );

```

## Handle presence

A client can respond to presence events generated by PubNub (client disconnection, ...) but can also send custom presence events to one or many channels.

```dart

_client.presence(['Test-Channel1'], {'state': 'busy'});

```

## Subscribe to PubNub events

In order to build a chat service for example, the client does not only send messages but must also consume messages, be aware if the client on the other end disconnected, status.

``` dart

_client.onStatusReceived.listen((status) => print('Status:${status.toString()}'));

_client.onPresenceReceived.listen((presence) => print('Presence:${presence.toString()}'));

_client.onMessageReceived.listen((message) => print('Message:$message'));

_client.onErrorReceived.listen((error) => print('Error:$error'));

_client.onMessageActionReceived.listen((action) => print('Action:$action'));

```

## Channel Groups

Channel groups access are handled by the PubNub Access Manager.

### Adding channels to a channel group



``` dart

_client.addChannelsToChannelGroup('Group1', ['Channel', 'Channel2']);

```

### List channels belonging to a channel group

``` dart

_client.listChannelsForChannelGroup('Group1').then((channels) {
                          print("Channels in Group 1: $channels");
                        });
                        
```

### Remove channels from a channel group

``` dart

 _client.removeChannelsFromChannelGroup('Group1', ['Channel']);
                        
```

### Delete a channel group

``` dart

_client.deleteChannelGroup('Group1');

```

### Subscribe to channel groups

``` dart

_client.subscribeToChannelGroups(['Group1']);

```

### Unsubscribe from channel groups

``` dart

_client.unsubscribeFromChannelGroups(['Group1']);

```

## Retrieve history

``` dart

_client.history('Channel', 1).then((items) {
                          if (items != null && items.isNotEmpty) {
                            print("Last Item: $items");
                          } else {
                            print('No items');
                          }
                        });
                        
```

## Push Notifications

Adding channels for push notifications using a Firebase messaging plugin push token

``` dart

_firebaseMessaging.getToken().then((token) {
    print("Token: $token");
    _client.addPushNotificationsOnChannels(PushType.FCM, token, ['Channel']);
    });
},

```

Listing channels enrolled in push notifications.

``` dart

_client.listPushNotificationChannels(PushType.FCM, token).then((channels) {
    print("Push Notes Channels: $channels");
});

```

Removing channels from push notifications.

``` dart

_client.removePushNotificationsFromChannels(PushType.FCM, token, ['Channel']);

```

Removing all channels from push notifications.

``` dart

_client.removeAllPushNotificationsFromDeviceWithPushToken(PushType.FCM, token);

```

## Signals

Signals allow to send small payloads (30 bytes max) in a very efficient and cost effective way.

``` dart

_client.signal(['Channel2'], {'signal': 'Hello Signal'});

```
## Message Actions

Message actions allow to receive message actions as well as adding some.
Each published message in pubNub has an associated time token. The publish method of the present plugin is returning a timeToken attribute longside others. Such attribute must be used when adding message actions, defined by a type and a value.

For example, if you would like to mark a message as received, you could pass the channel, the time token of the original message, a type, for example 'receipt' and a value, for example 'received' or 'read' if you want to indicate tht the message was read as iMessage does on iPhone.

``` dart
_client.addMessageAction(['Channel2'], 42154216241, 'receipt', 'received');
```

In order to listen for message actions on messages, it is as easy as registering for message action events:

``` dart
_client.onMessageActionReceived.listen((action) => print('Action:$action'));
```

## Cleanup

When done with a PubNub client, the client can be closed and cleaned up.

```dart

_client.dispose();

```

## License

This plugin is under an Apache 2.0 license

## Features and bugs

Please file feature requests and bugs at the issue tracker. https://github.com/Ingenio/flutter_pubnub/issues


CHANGELOG.md
## [0.0.1]

* Android and iOS PubNub Flutter plugin. Implements subscribe, unsubscribe, publish, status, filter expressions, presence and error handling. Also allows multiple PubNub clients to be created.

## [0.1.0]

* Refactored code so spread operators are not used

## [0.1.1]

* Documentation tweaks

## [0.1.2]

* Breaking changes: Now publish, presence and unsubscribe take an array of channels

## [0.1.3]

* Removed unused constant and documentation tweaks

## [0.1.4]

* Added channel groups handling, push notifications and history

## [0.2.0]

* Added PubNub signals

## [0.3.0]

* Fixed crashes caused by code obfuscation. 
* Added Proguard rules

## [0.3.1]

* File format warning removed

## [0.3.2]

* Updated PubNub dependency to latest (4.12.0)

## [0.4.0]

* Update PubNub dependency to latest Android SDK 4.32, solving network related issues

## [0.4.1]

* Fixed typo on Android version: 4.31

## [0.4.2]

* Changed minimum version for iOS to 8.0

## [0.5.0]

* Updated to iOS PubNub 4.13.1

## [1.0.0]

* Updated to null safety
* Added message actions
* Bug fixes and other improvements


example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:flutter_pubnub_example/main.dart';

void main() {
  testWidgets('Verify Platform version', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(MyApp());

    // Verify that platform version is retrieved.
    expect(
      find.byWidgetPredicate(
        (Widget widget) => widget is Text &&
                           widget.data.startsWith('Running on:'),
      ),
      findsOneWidget,
    );
  });
}


lib/generated/i18n.dart
import 'dart:async';

import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';

// ignore_for_file: non_constant_identifier_names
// ignore_for_file: camel_case_types
// ignore_for_file: prefer_single_quotes

// This file is automatically generated. DO NOT EDIT, all your changes would be lost.
class S implements WidgetsLocalizations {
  const S();

  static S current;

  static const GeneratedLocalizationsDelegate delegate = GeneratedLocalizationsDelegate();

  static S of(BuildContext context) => Localizations.of<S>(context, S);

  @override
  TextDirection get textDirection => TextDirection.ltr;
}

class $en extends S {
  const $en();
}

class GeneratedLocalizationsDelegate extends LocalizationsDelegate<S> {
  const GeneratedLocalizationsDelegate();

  List<Locale> get supportedLocales {
    return const <Locale>[
      Locale("en", ""),
    ];
  }

  LocaleListResolutionCallback listResolution({Locale fallback, bool withCountry = true}) {
    return (List<Locale> locales, Iterable<Locale> supported) {
      if (locales == null || locales.isEmpty) {
        return fallback ?? supported.first;
      } else {
        return _resolve(locales.first, fallback, supported, withCountry);
      }
    };
  }

  LocaleResolutionCallback resolution({Locale fallback, bool withCountry = true}) {
    return (Locale locale, Iterable<Locale> supported) {
      return _resolve(locale, fallback, supported, withCountry);
    };
  }

  @override
  Future<S> load(Locale locale) {
    final String lang = getLang(locale);
    if (lang != null) {
      switch (lang) {
        case "en":
          S.current = const $en();
          return SynchronousFuture<S>(S.current);
        default:
        // NO-OP.
      }
    }
    S.current = const S();
    return SynchronousFuture<S>(S.current);
  }

  @override
  bool isSupported(Locale locale) => _isSupported(locale, true);

  @override
  bool shouldReload(GeneratedLocalizationsDelegate old) => false;

  ///
  /// Internal method to resolve a locale from a list of locales.
  ///
  Locale _resolve(Locale locale, Locale fallback, Iterable<Locale> supported, bool withCountry) {
    if (locale == null || !_isSupported(locale, withCountry)) {
      return fallback ?? supported.first;
    }

    final Locale languageLocale = Locale(locale.languageCode, "");
    if (supported.contains(locale)) {
      return locale;
    } else if (supported.contains(languageLocale)) {
      return languageLocale;
    } else {
      final Locale fallbackLocale = fallback ?? supported.first;
      return fallbackLocale;
    }
  }

  ///
  /// Returns true if the specified locale is supported, false otherwise.
  ///
  bool _isSupported(Locale locale, bool withCountry) {
    if (locale != null) {
      for (Locale supportedLocale in supportedLocales) {
        // Language must always match both locales.
        if (supportedLocale.languageCode != locale.languageCode) {
          continue;
        }

        // If country code matches, return this locale.
        if (supportedLocale.countryCode == locale.countryCode) {
          return true;
        }

        // If no country requirement is requested, check if this locale has no country.
        if (true != withCountry && (supportedLocale.countryCode == null || supportedLocale.countryCode.isEmpty)) {
          return true;
        }
      }
    }
    return false;
  }
}

String getLang(Locale l) =>
    l == null ? null : l.countryCode != null && l.countryCode.isEmpty ? l.languageCode : l.toString();


ios/Classes/PubnubPlugin.h
#import <Flutter/Flutter.h>

#import <Flutter/Flutter.h>
#import <PubNub/PubNub.h>

@class MessageStreamHandler;
@class StatusStreamHandler;
@class ErrorStreamHandler;
@class PresenceStreamHandler;
@class MessageActionStreamHandler;

@interface PubnubPlugin : NSObject<FlutterPlugin>
@property (nonatomic, strong) MessageStreamHandler *messageStreamHandler;
@property (nonatomic, strong) StatusStreamHandler *statusStreamHandler;
@property (nonatomic, strong) PresenceStreamHandler *presenceStreamHandler;
@property (nonatomic, strong) ErrorStreamHandler *errorStreamHandler;
@property (nonatomic, strong) MessageActionStreamHandler *messageActionStreamHandler;
@end

@interface MessageStreamHandler : NSObject<FlutterStreamHandler>
@property (nonatomic, strong) FlutterEventSink eventSink;

- (void) sendMessage:(PNMessageResult *)message clientId:(NSString *)clientId;
- (void) sendSignal:(PNSignalResult *)signal clientId:(NSString *)clientId;

@end

@interface StatusStreamHandler : NSObject <FlutterStreamHandler>
@property (nonatomic, strong) FlutterEventSink eventSink;

- (void) sendStatus:(PNStatus *)status clientId:(NSString *)clientId;

@end

@interface PresenceStreamHandler : NSObject <FlutterStreamHandler>
@property (nonatomic, strong) FlutterEventSink eventSink;

- (void) sendPresence:(PNPresenceEventResult *)presence clientId:(NSString *)clientId;

@end

@interface ErrorStreamHandler : NSObject <FlutterStreamHandler>
@property (nonatomic, strong) FlutterEventSink eventSink;

- (void) sendError:(NSDictionary *)error;

@end

@interface MessageActionStreamHandler : NSObject <FlutterStreamHandler>
@property (nonatomic, strong) FlutterEventSink eventSink;

- (void) sendMessageAction:(PNMessageActionResult *)action clientId:(NSString *)clientId;

@end

@interface MissingArgumentException : NSException
@end



example/lib/main.dart
import 'dart:io';

import 'package:dio/dio.dart';
import 'package:firebase_core/firebase_core.dart';
import 'package:firebase_messaging/firebase_messaging.dart';
import 'package:flutter/material.dart';
import 'package:flutter_pubnub/pubnub.dart';

void main() async {
  await Firebase.initializeApp();
  runApp(MyApp());
}

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  final _firstUserClient = PubNub(PubNubConfig('pub-c-xxx', 'sub-c-xxx',
      uuid: 'a0a80f2d-b48d-460c-b3bd-a244a877df1f'));
  final _secondUserClient = PubNub(PubNubConfig('pub-c-zzz', 'sub-c-zzz',
      presenceTimeout: 120,
      uuid: '127c1ab5-fc7f-4c46-8460-3207b6782007',
      filter: 'uuid != "127c1ab5-fc7f-4c46-8460-3207b6782007"'));

  final FirebaseMessaging _firebaseMessaging = FirebaseMessaging.instance;

  @override
  void initState() {
    super.initState();
    _firstUserClient.uuid().then((uuid) {
      print('UUID1: $uuid');
      sendEvent('UUID', uuid);
    });
    _secondUserClient.uuid().then((uuid) => print('UUID2: $uuid'));
    _firstUserClient.onStatusReceived.listen((status) {
      print('Status:${status.toString()}');
      sendEvent('STATUS', status.toString());
    });
    _firstUserClient.onPresenceReceived.listen((presence) {
      print('Presence:${presence.toString()}');
      sendEvent('PRESENCE', presence.toString());
    });
    _firstUserClient.onMessageReceived.listen((message) {
      print('Message:$message');
      sendEvent('MESSAGE', message.toString());
    });

    _firstUserClient.onErrorReceived.listen((error) {
      print('Error:$error');
      sendEvent('ERROR', error.toString());
    });

    if (Platform.isIOS) iOSPermission();

    FirebaseMessaging.onMessage.listen((RemoteMessage message) {
      print("onMessage: $message");
      sendEvent('PUSH MESSAGE', message.toString());
    });

    FirebaseMessaging.onMessageOpenedApp.listen((RemoteMessage message) {
      print("onMessageOpenedApp: $message");
    });
  }

  void iOSPermission() async {
    NotificationSettings settings = await _firebaseMessaging.requestPermission(
      alert: true,
      badge: true,
      sound: true,
    );
    print("Settings registered: $settings");
  }

  Future sendEvent(String eventName, String message) async {
    Dio dio = new Dio(); // with default Options

    await dio.post("http://logs-01.loggly.com/inputs/<loggly_key>/tag/http",
        data: {"flutter_pubnub": eventName, "message": message});
  }

  @override
  void dispose() {
    print('Unsubscribe all');
    _firstUserClient.unsubscribeAll();
    _secondUserClient.unsubscribeAll();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) => MaterialApp(
        home: Scaffold(
          appBar: AppBar(
            title: const Text('PubNub'),
          ),
          body: Column(mainAxisAlignment: MainAxisAlignment.center, children: <
              Widget>[
            Padding(
                padding: EdgeInsets.only(left: 8.0),
                child: Row(
                    mainAxisAlignment: MainAxisAlignment.start,
                    children: <Widget>[Text('Pub/Sub/Publish')])),
            Row(
                mainAxisAlignment: MainAxisAlignment.spaceAround,
                children: <Widget>[
                  FlatButton(
                      color: Colors.black12,
                      onPressed: () {
                        _firstUserClient.unsubscribe(['Channel']);
                        _secondUserClient.unsubscribe(['Channel']);
                      },
                      child: Text('Unsubscribe')),
                  FlatButton(
                      color: Colors.black12,
                      onPressed: () {
                        _firstUserClient.subscribe(['Channel', 'Channel2']);
                        _secondUserClient.subscribe(['Channel']);
                      },
                      child: Text('Subscribe')),
                  FlatButton(
                      color: Colors.black12,
                      onPressed: () {
                        _firstUserClient.publish(['Channel', 'Channel2'],
                            {'message': 'Hello World!'});
                        //_secondUserClient.publish(['Channel'], {'message': 'Hello First User!'},
                        //   metadata: {'uuid': '127c1ab5-fc7f-4c46-8460-3207b6782007'});
                        // _firstUserClient.presence(['Channel'], {'state': 'AFK'});
                      },
                      child: Text('Send Message')),
                ]),
            Padding(
                padding: EdgeInsets.all(
                  8.0,
                ),
                child: Divider(
                  height: 1,
                  color: Colors.black,
                )),
            Padding(
                padding: EdgeInsets.only(left: 8.0),
                child: Row(
                    mainAxisAlignment: MainAxisAlignment.start,
                    children: <Widget>[Text('Channel Group')])),
            Row(
                mainAxisAlignment: MainAxisAlignment.spaceAround,
                children: <Widget>[
                  FlatButton(
                      color: Colors.black12,
                      onPressed: () {
                        _firstUserClient.addChannelsToChannelGroup(
                            'Group1', ['Channel', 'Channel2']);
                      },
                      child: Text('Add')),
                  FlatButton(
                      color: Colors.black12,
                      onPressed: () {
                        _firstUserClient
                            .listChannelsForChannelGroup('Group1')
                            .then((channels) {
                          print("Channels in Group 1: $channels");
                        });
                      },
                      child: Text('List')),
                  FlatButton(
                      color: Colors.black12,
                      onPressed: () {
                        _firstUserClient.removeChannelsFromChannelGroup(
                            'Group1', ['Channel']).then((void arg) {
                          _firstUserClient
                              .listChannelsForChannelGroup('Group1')
                              .then((channels) {
                            print(
                                "Channels in Group 1 after deletion: $channels");
                          });
                        });
                      },
                      child: Text('Remove')),
                ]),
            Row(
                mainAxisAlignment: MainAxisAlignment.spaceAround,
                children: <Widget>[
                  FlatButton(
                      color: Colors.black12,
                      onPressed: () {
                        _firstUserClient.deleteChannelGroup('Group1');
                      },
                      child: Text('Delete')),
                  FlatButton(
                      color: Colors.black12,
                      onPressed: () {
                        _firstUserClient.subscribeToChannelGroups(['Group1']);
                      },
                      child: Text('Subscribe')),
                  FlatButton(
                      color: Colors.black12,
                      onPressed: () {
                        _firstUserClient
                            .unsubscribeFromChannelGroups(['Group1']);
                      },
                      child: Text('Unsubscribe')),
                ]),
            Padding(
                padding: EdgeInsets.all(
                  8.0,
                ),
                child: Divider(
                  height: 1,
                  color: Colors.black,
                )),
            Padding(
                padding: EdgeInsets.only(left: 8.0),
                child: Row(
                    mainAxisAlignment: MainAxisAlignment.start,
                    children: <Widget>[Text('History')])),
            Row(
                mainAxisAlignment: MainAxisAlignment.spaceAround,
                children: <Widget>[
                  FlatButton(
                      color: Colors.black12,
                      onPressed: () {
                        _firstUserClient.history('Channel', 1).then((items) {
                          if (items != null && items.isNotEmpty) {
                            print("Last Item: $items");
                          } else {
                            print('No items');
                          }
                        });
                      },
                      child: Text('Last')),
                ]),
            Padding(
                padding: EdgeInsets.all(
                  8.0,
                ),
                child: Divider(
                  height: 1,
                  color: Colors.black,
                )),
            Padding(
                padding: EdgeInsets.only(left: 8.0),
                child: Row(
                    mainAxisAlignment: MainAxisAlignment.start,
                    children: <Widget>[Text('Push Notifications')])),
            Row(
                mainAxisAlignment: MainAxisAlignment.spaceAround,
                children: <Widget>[
                  FlatButton(
                      color: Colors.black12,
                      onPressed: () {
                        _firebaseMessaging.getToken().then((token) {
                          print("Token: $token");
                          _firstUserClient.addPushNotificationsOnChannels(
                              PushType.FCM, token!, ['Channel']);
                          sendEvent(
                              'ADD CHANNELS', ['Channel'].toList().toString());
                        });
                      },
                      child: Text('Add')),
                  FlatButton(
                      color: Colors.black12,
                      onPressed: () {
                        _firebaseMessaging.getToken().then((token) {
                          print("Token: $token");
                          _firstUserClient
                              .listPushNotificationChannels(PushType.FCM, token!)
                              .then((channels) {
                            print("Push Notes Channels: $channels");
                            sendEvent(
                                'LIST CHANNELS', channels.toList().toString());
                          });
                        });
                      },
                      child: Text('List')),
                  FlatButton(
                      color: Colors.black12,
                      onPressed: () {
                        _firebaseMessaging.getToken().then((token) {
                          print("Token: $token");
                          _firstUserClient.removePushNotificationsFromChannels(
                              PushType.FCM, token!, ['Channel']);
                          sendEvent('REMOVE CHANNELS',
                              ['Channel'].toList().toString());
                        });
                      },
                      child: Text('Remove')),
                  FlatButton(
                      color: Colors.black12,
                      onPressed: () {
                        _firebaseMessaging.getToken().then((token) {
                          print("Token: $token");
                          _firstUserClient
                              .removeAllPushNotificationsFromDeviceWithPushToken(
                                  PushType.FCM, token!);
                          sendEvent('REMOVE ALL CHANNELS', token!);
                        });
                      },
                      child: Text('Remove All')),
                ]),
            Row(
                mainAxisAlignment: MainAxisAlignment.spaceAround,
                children: <Widget>[
                  FlatButton(
                      color: Colors.black12,
                      onPressed: () {
                        _firebaseMessaging.getToken().then((token) {
                          print("Token: $token");
                          sendEvent('TOKEN', token!);
                        });
                      },
                      child: Text('Token')),
                ]),
            Padding(
                padding: EdgeInsets.all(
                  8.0,
                ),
                child: Divider(
                  height: 1,
                  color: Colors.black,
                )),
            Padding(
                padding: EdgeInsets.only(left: 8.0),
                child: Row(
                    mainAxisAlignment: MainAxisAlignment.start,
                    children: <Widget>[Text('Signals')])),
            Row(
                mainAxisAlignment: MainAxisAlignment.spaceAround,
                children: <Widget>[
                  FlatButton(
                      color: Colors.black12,
                      onPressed: () {
                        _firstUserClient
                            .signal(['Channel2'], {'signal': 'Hello Signal'});
                      },
                      child: Text('Signal')),
                ]),
          ]),
        ),
      );
}
